`Multiple GeneratedClassBuildItem were produced for the same classes` when using abstract class

### Describe the bug

Hi, I spot in one of our test module when using the abstract class the follow up error is present
```
[ERROR]         [error]: Build step io.quarkus.deployment.pkg.steps.JarResultBuildStep#buildRunnerJar threw an exception: java.lang.IllegalStateException: Multiple GeneratedClassBuildItem were produced for the same classes:
[ERROR] 
[ERROR] - org.acme.tracable.TraceableResource$quarkusrestinvoker$getLastTraceId_3736e02108c2d04115908baa7a205acb1fee7076: 2
```

It probably introduced by https://github.com/quarkusio/quarkus/pull/50278.

We have one abstract class defining endpoint. This class  is used on SSE ping and pong resources (For testing we extend even grpc resource by it, for it's not present in reproducer).

I try to remove the endpoint annotation from affected method and it's possible to build it.

For more info see the reproducer. 


### Expected behavior

Build the jar without errors

### Actual behavior
```
[ERROR]         [error]: Build step io.quarkus.deployment.pkg.steps.JarResultBuildStep#buildRunnerJar threw an exception: java.lang.IllegalStateException: Multiple GeneratedClassBuildItem were produced for the same classes:
[ERROR] 
[ERROR] - org.acme.tracable.TraceableResource$quarkusrestinvoker$getLastTraceId_3736e02108c2d04115908baa7a205acb1fee7076: 2
```

### How to Reproduce?

1. `git clone -b multple-generated-items git@github.com:jedla97/quarkus-reproducers.git`
2. `cd quarkus-reproducers`
3. `mvn clean package -DskipTests`
4. The error is there

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

21.0.2+13-58

### Quarkus version or git rev

main

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9 (8e8579a9e76f7d015ee5ec7bfcdc97d260186937)

### Additional information

_No response_

--- a/extensions/resteasy-reactive/rest/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/QuarkusInvokerFactory.java
+++ b/extensions/resteasy-reactive/rest/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/QuarkusInvokerFactory.java
@@ -5,7 +5,9 @@ import static org.jboss.jandex.gizmo2.Jandex2Gizmo.methodDescOf;
 import java.lang.reflect.Modifier;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.function.Predicate;
 import java.util.function.Supplier;
 
@@ -32,6 +34,8 @@ public class QuarkusInvokerFactory implements EndpointInvokerFactory {
     final BuildProducer<GeneratedClassBuildItem> generatedClassBuildItemBuildProducer;
     final ResteasyReactiveRecorder recorder;
 
+    private final Map<String, Supplier<EndpointInvoker>> generatedInvokers = new HashMap<>();
+
     public QuarkusInvokerFactory(Predicate<String> applicationClassPredicate,
             BuildProducer<GeneratedClassBuildItem> generatedClassBuildItemBuildProducer,
             ResteasyReactiveRecorder recorder) {
@@ -51,6 +55,9 @@ public class QuarkusInvokerFactory implements EndpointInvokerFactory {
 
         String baseName = currentClassInfo.name() + "$quarkusrestinvoker$" + method.getName() + "_"
                 + HashUtil.sha1(endpointIdentifier);
+        if (generatedInvokers.containsKey(baseName)) {
+            return generatedInvokers.get(baseName);
+        }
         ClassOutput classOutput = new GeneratedClassGizmo2Adaptor(generatedClassBuildItemBuildProducer, null,
                 applicationClassPredicate.test(currentClassInfo.name().toString()));
         Gizmo g = Gizmo.create(classOutput);
@@ -80,7 +87,9 @@ public class QuarkusInvokerFactory implements EndpointInvokerFactory {
                 });
             });
         });
-        return recorder.invoker(baseName);
+        var result = recorder.invoker(baseName);
+        generatedInvokers.put(baseName, result);
+        return result;
     }
 
 }

--- a/integration-tests/rest-client-reactive-http2/pom.xml
+++ b/integration-tests/rest-client-reactive-http2/pom.xml
@@ -43,6 +43,11 @@
             <artifactId>vertx-web-client</artifactId>
             <scope>test</scope>
         </dependency>
+      <dependency>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-junit5-internal</artifactId>
+        <scope>test</scope>
+      </dependency>
 
         <!-- Minimal test dependencies to *-deployment artifacts for consistent build order -->
         <dependency>
@@ -85,6 +90,30 @@
                         </goals>
                     </execution>
                 </executions>
+            </plugin>
+            <plugin>
+              <groupId>org.apache.maven.plugins</groupId>
+              <artifactId>maven-surefire-plugin</artifactId>
+              <configuration>
+                <runOrder>alphabetical</runOrder>
+              </configuration>
+              <executions>
+                <!--
+                The prod mode tests need to be part of a different execution to ensure that they don't mess with the standard tests.
+                By adding this configuration we ensure that the maven surefire plugin will execute twice, one for the regular **/*Test.java
+                tests (using the 'default-test' execution), and one for the prod mode tests (this 'prod-mode' execution)
+                -->
+                <execution>
+                  <id>prod-mode</id>
+                  <phase>test</phase>
+                  <goals>
+                    <goal>test</goal>
+                  </goals>
+                  <configuration>
+                    <includes>**/*PMT.java</includes>
+                  </configuration>
+                </execution>
+              </executions>
             </plugin>
         </plugins>
     </build>

--- a/integration-tests/rest-client-reactive-http2/src/test/java/io/quarkus/it/rest/client/http2/RestMethodInAbstractPMT.java
+++ b/integration-tests/rest-client-reactive-http2/src/test/java/io/quarkus/it/rest/client/http2/RestMethodInAbstractPMT.java
@@ -1,0 +1,75 @@
+package io.quarkus.it.rest.client.http2;
+
+import static io.restassured.RestAssured.get;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusProdModeTest;
+
+/**
+ * This test does not really belong here, but the only way to create a test for
+ * <a href="https://github.com/quarkusio/quarkus/issues/50314">this issue</a>
+ * is to build a similar REST application in prod mode.
+ */
+public class RestMethodInAbstractPMT {
+
+    @RegisterExtension
+    static QuarkusProdModeTest runner = new QuarkusProdModeTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(BaseResource.class, Resource1.class, Resource2.class, Client.class))
+            .setRun(true);
+
+    @Test
+    public void test() {
+        get("/res1")
+                .then()
+                .statusCode(200)
+                .body(Matchers.equalTo("foo"));
+    }
+
+    public static abstract class BaseResource {
+
+        @GET
+        @Path("/base")
+        @Produces(MediaType.TEXT_PLAIN)
+        public String base() {
+            return "base";
+        }
+    }
+
+    @Path("res1")
+    public static class Resource1 extends BaseResource {
+
+        @GET
+        public String foo() {
+            return "foo";
+        }
+    }
+
+    @Path("res2")
+    public static class Resource2 extends BaseResource {
+
+        @RestClient
+        Client client;
+
+        @GET
+        public String foo() {
+            return "foo";
+        }
+
+        @GET
+        @Path("bar")
+        public String bar() {
+            return "bar";
+        }
+    }
+
+}

